registry/storage/layercache.go

```diff
@@ -0,0 +1,232 @@
+// available tells the caller that the layer exists in the repository used by
+// this layerInfoCache. This is used as an access check before looking up
+// global path information. If false is returned, the caller should still
+// check the backend to if it exists elsewhere.
+func (lc *layerInfoCache) available(conn redis.Conn, dgst digest.Digest) (bool, error) {
+	logrus.Infof("(*layerInfoCache).available(%q)", dgst)
+	return redis.Bool(conn.Do("SISMEMBER", lc.repositoryBlobsKey(), dgst))
+}
+
+// add the layer to the cache.
+func (lc *layerInfoCache) add(conn redis.Conn, layer distribution.Layer) error {
+	// first, add the raw blob data
+	if err := lc.addLayer(conn, layer); err != nil {
+		return err
+	}
+
+	// second, add the layer to the repo
+	return lc.addDigest(conn, layer.Digest())
+}
+
+// addLayer stores the layer info in the cache.
+func (lc *layerInfoCache) addLayer(conn redis.Conn, layer distribution.Layer) error {
+	path, err := lc.resolveLayerPath(layer)
+	if err != nil {
+		return err
+	}
+	// Store the data in a hash. Using a hash here since we might store
+	// other layer related fields in the future.
+	_, err = conn.Do("HMSET", blobKey(layer.Digest()), "path", path, "length", layer.Length())
+	return err
+}
+
+// addDigest associates the digest with the repository set.
+func (lc *layerInfoCache) addDigest(conn redis.Conn, dgst digest.Digest) error {
+	// now, associate it with the repository.
+	_, err := conn.Do("SADD", lc.repositoryBlobsKey(), dgst)
+	return err
+}
+
+// fetch grabs the layer from the cache, if available.
+func (lc *layerInfoCache) fetch(conn redis.Conn, dgst digest.Digest) (distribution.Layer, error) {
+	logrus.Infof("(*layerInfoCache).fetch(%q)", dgst)
+	reply, err := redis.Values(conn.Do("HMGET", blobKey(dgst), "path", "length"))
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		path   string
+		length int64
+	)
+
+	if _, err := redis.Scan(reply, &path, &length); err != nil {
+		return nil, err
+	}
+
+	if path == "" {
+		return nil, distribution.ErrUnknownLayer{
+			FSLayer: manifest.FSLayer{BlobSum: dgst},
+		}
+	}
+
+	return newLayerReader(lc.driver, dgst, path, length)
+}
+
+// extractLayerInfo pulls the layerInfo from the layer, attempting to get the
+// path information from either the concrete object or by resolving the
+// primary blob store path.
+func (lc *layerInfoCache) resolveLayerPath(layer distribution.Layer) (path string, err error) {
+	// try and resolve the type and driver, so we don't have to traverse links
+	switch v := layer.(type) {
+	case *layerReader:
+		// only set path if we have same driver instance.
+		if v.driver == lc.driver {
+			return v.path, nil
+		}
+	}
+
+	logrus.Warnf("resolving layer path during cache lookup (%v@%v)", lc.Repository.Name(), layer.Digest())
+	// we have to do an expensive stat to resolve the layer location but no
+	// need to check the link, since we already have layer instance for this
+	// repository.
+	bp, err := lc.blobStore.path(layer.Digest())
```

Comment:
Yes, effectively, we are caching the membership to the repository and the blobs primary location in the backend.